### ISSUE TITLE
How to use embedded lua interpreter for scriptlets?

### ISSUE THREAD

**Issue by moedan** [ctron/rpm-builder#101]
Hi,

I've problems getting the embedded lua interpreter of rpm to work with this plugin.

According to the [documentation of this plugin](https://ctron.github.io/rpm-builder/scripts.html#basic-configuration), I have used the following configuration:

```
<configuration>
    <packageName>...</packageName>
    <afterTransaction>
        <interpreter>&lt;lua&gt;</interpreter>
        <file>${project.basedir}/src/main/packaging/posttrans.lua</file>
    </afterTransaction>
    <entries>
        ...
    </entries>
</configuration>
```

Content of **posttrans.lua**:
```
print("Hello, World!")
```

Unfortunately, I can't install the package (used container image with Fedora 40). Instead, I get the following error message:
```
$ dnf install /tmp/mwe-1.0-0.*.noarch.rpm
Fedora 40 - x86_64
Fedora 40 openh264 (From Cisco) - x86_64
Fedora 40 - x86_64 - Updates
Error: 
 Problem: conflicting requests
  - nothing provides <lua> needed by mwe-1.0-0.202408121229.noarch from @commandline
(try to add '--skip-broken' to skip uninstallable packages)
```

However, the installation of another package with a lua posttrans scriptlet (e.g. 'glibc-all-langpacks') works without any problems.

When I check the generated RPM package, I get the expected output (as for example with the package 'glibc-all-langpacks'):
```
$ rpm -q --queryformat="%{NAME}: posttrans interpreter: %{POSTTRANSPROG}\n" glibc-all-langpacks /tmp/mwe-1.0-0.*.noarch.rpm
glibc-all-langpacks: posttrans interpreter: <lua>
mwe: posttrans interpreter: <lua>
```

I've added an MWE ([mwe.zip](https://github.com/user-attachments/files/16582752/mwe.zip)) including a Containerfile to demonstrate the problem.
Steps to reproduce:

```
# In root directory:
mvn clean package
podman build -t embeddedluatest .
podman run -it embeddedluatest

# In container:
dnf install /tmp/mwe-1.0-0.*.noarch.rpm
```

Do I have the interpreter configured correctly? Have I missed something else?
Thanks in advance!

**Comment by ctron** [comment]
I've never done that. Looks interesting :) … I'd assume that you need a full path, like `/usr/bin/lua` (or wherever that binaries lives).

**Comment by moedan** [comment]
Thanks for your response!

From my understanding, the path `<lua>` is the correct syntax to call the internal Lua interpreter of the rpm-command. 
In the [official documentation of this rpm feature](https://rpm-software-management.github.io/rpm/manual/lua), the following example is given for the use of the internal Lua interpreter in the spec file:
```
%pre -p <lua>
print('Hello from Lua')
```

Apart from lua-libs, there are no Lua binaries installed in my example container with Fedora 40:

```
$ dnf list installed lua*
Installed Packages
lua-libs.x86_64               5.4.6-5.fc40                @4a49306984b248a0bf3a4434ca6697b2
```

Nevertheless, I can evaluate Lua commands with the rpm command:

```
$ rpm --eval '%{lua:print("hello")}'
hello
```

<hr/>

I have further analysed the differences between the package `mwe` (generated by the rpm-builder plugin) and the package `glibc-all-langpacks`. 
Here I noticed that a ‘reqired’ entry for `<lua>` was added in the package generated by the rpm-builder plugin: 

```
$ rpm -q --requires mwe-1.0-0.202408121345.noarch.rpm 
<lua>
<lua>
rpmlib(CompressedFileNames) <= 3.0.4-1
rpmlib(FileDigests) <= 4.6.0-1
rpmlib(PayloadFilesHavePrefix) <= 4.0-1
```

However, the entry is not present in the `glibc-all-langpacks` package:

```
$ rpm -q --requires glibc-all-langpacks-2.39-22.fc40.x86_64.rpm 
glibc = 2.39-22.fc40
glibc-common = 2.39-22.fc40
rpmlib(BuiltinLuaScripts) <= 4.2.2-1
rpmlib(CompressedFileNames) <= 3.0.4-1
rpmlib(FileDigests) <= 4.6.0-1
rpmlib(PartialHardlinkSets) <= 4.0.4-1
rpmlib(PayloadFilesHavePrefix) <= 4.0-1
rpmlib(PayloadIsZstd) <= 5.4.18-1
rpmlib(ScriptletExpansion) <= 4.9.0-1
```

As a test, I rebuilt the `mwe` package with the command `rpmrebuild` and removed the 'requird' entry for `<rpm>`. 

```
$ rpmrebuild --package --notest-install -e mwe-1.0-0.202408121345.noarch.rpm 
```

Interestingly, the installation of the rebuilt package then works with the expected output:

```
$ dnf install /root/rpmbuild/RPMS/noarch/mwe-1.0-0.202408121345.noarch.rpm 
Last metadata expiration check: 0:46:27 ago on Mon Aug 26 09:11:46 2024.
Dependencies resolved.
==========================================================================================================================================================================
 Package                           Architecture                         Version                                          Repository                                  Size
==========================================================================================================================================================================
Installing:
 mwe                               noarch                               1.0-0.202408121345                               @commandline                               6.6 k

Transaction Summary
==========================================================================================================================================================================
Install  1 Package

Total size: 6.6 k
Installed size: 11  
Is this ok [y/N]: y
Downloading Packages:
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Preparing        :                                                                                                                                                  1/1 
  Installing       : mwe-1.0-0.202408121345.noarch                                                                                                                    1/1 
  Running scriptlet: mwe-1.0-0.202408121345.noarch                                                                                                                    1/1 
Hello, World!

Installed:
  mwe-1.0-0.202408121345.noarch                                                                                                                                           

Complete!
```

Do you know why the required entry `<lua>` is added to the rpm build by the rpm-build plugin?

**Comment by ctron** [comment]
Hm ok … what happens is, when using a script, that the script interpreter is added as a requirement, by using a file name.

If you're using an internal interpreter, then that might not work. I assume it would need `rpmlib(BuiltinLuaScripts) <= 4.2.2-1` instead in this case.

**Comment by moedan** [comment]
Thank you for your explanation!

It looks to me as if this feature requires a change in the [RpmBuilder](https://github.com/eclipse/packager/blob/50e4a5559a78bba9ed3495e10bb125db5f0cef7c/rpm/src/main/java/org/eclipse/packager/rpm/build/RpmBuilder.java#L1180) (line 1180ff) of [Eclipse Packager project](https://github.com/eclipse/packager).

As a proof of concept, I added a simple check for the internal Lua interpreter in the Eclipse Packager RPM library. In case of the internal Lua interpeter I added the dependency `rpmlib(BuiltinLuaScripts) <= 4.2.2-1` as you suggested. With this 'fix' was able to install the generated rpm package in the container.

My proof of concept for `RpmBuilder::setPostTransactionScript`:

```java
private static final String INTERNAL_LUA_INTERPRETER = "<lua>";

public void setPostTransactionScript(final String interpreter, final String script) {
    if (Objects.equals(interpreter, INTERNAL_LUA_INTERPRETER)) {
        addRequirement("rpmlib(BuiltinLuaScripts)", "4.2.2-1", RpmDependencyFlags.INTERPRETER, RpmDependencyFlags.LESS, RpmDependencyFlags.EQUAL, RpmDependencyFlags.RPMLIB);
        addRequirement("rpmlib(BuiltinLuaScripts)", "4.2.2-1", RpmDependencyFlags.SCRIPT_POSTUN, RpmDependencyFlags.INTERPRETER, RpmDependencyFlags.LESS, RpmDependencyFlags.EQUAL, RpmDependencyFlags.RPMLIB);
    } else {
        addRequirement(interpreter, null, RpmDependencyFlags.INTERPRETER);
        addRequirement(interpreter, null, RpmDependencyFlags.SCRIPT_POSTUN, RpmDependencyFlags.INTERPRETER);
    }
    setScript(RpmTag.POSTTRANSACTION_SCRIPT_PROG, RpmTag.POSTTRANSACTION_SCRIPT, interpreter, script);
}

```

**Comment by ctron** [comment]
Looks good. The question would be: how to handle the version? Make that configurable with a reasonable default?

**Comment by dwalluck** [comment]
The `rpmlib(BuiltinLuaScripts) <= 4.2.2-1` should be a fixed version, with 4.2.2 being the RPM version.

**Comment by moedan** [comment]
The embedded lua interpreter of rpm is now supported for scriptlets. :tada:
It can be configured as described in the [opening post](https://github.com/ctron/rpm-builder/issues/101#issue-2461059278) by defining `<lua>` as interpreter.

Thanks for your support! :+1:
